### PR TITLE
feat: add `@npmcli/package-json` package

### PR DESCRIPTION
**Pull request by MichaelDeBoey**
Please fill in this template.

- [x] Use a meaningful title for the pull request. Include the name of the package modified.
- [x] Test the change in your own code. (Compile and run.)
- [x] [Add or edit tests](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#my-package-teststs) to reflect the change.
- [x] Follow the advice from the [readme](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#make-a-pull-request).
- [x] Avoid [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#common-mistakes).
- [x] [Run `npm test <package to test>`](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#running-tests).

Select one of these and delete the others:

If adding a new definition:
- [x] The package does not already provide its own types, or cannot have its `.d.ts` files generated via `--declaration`
- [x] If this is for an npm package, match the name. If not, do not conflict with the name of an npm package.
- [x] Create it with `dts-gen --dt`, not by basing it on an existing project.
- [x] Represents shape of module/library [correctly](https://www.typescriptlang.org/docs/handbook/declaration-files/library-structures.html)
- [x] `tslint.json` [should contain](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#linter-tslintjson) `{ "extends": "@definitelytyped/dtslint/dt.json" }`, and no additional rules.
- [x] `tsconfig.json` [should have](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#tsconfigjson) `noImplicitAny`, `noImplicitThis`, `strictNullChecks`, and `strictFunctionTypes` set to `true`.

### CODE REVIEW

**Comment by MichaelDeBoey** [comment]
Tests will be fixed once https://github.com/microsoft/DefinitelyTyped-tools/pull/429 is merged

**Review by andrewbranch — CHANGES_REQUESTED** [review]
I don’t think we should be bringing in all of type-fest to get one PackageJson type here.

**Comment by MichaelDeBoey** [comment]
@andrewbranch copy/pasting the `PackageJson` type doesn't seem like a good idea, so what would be your proposal then?

**Comment by andrewbranch** [comment]
Copy/pasting a well-authored type from somewhere else, as licensing allows, seems ok to me. DT consumer UX >>> DT maintainer DX in importance. (I also wonder if this type is appropriate for the `content` property at all, as my assumption is this library will return whatever object is in a package.json file, even if it violates the expected schema.)

**Comment by MichaelDeBoey** [comment]
> I also wonder if this type is appropriate for the `content` property at all, as my assumption is this library will return whatever object is in a package.json file, even if it violates the expected schema.

@andrewbranch `@npmcli/package-json` is indeed just taking whatever object is in `package.json`. 
https://github.com/npm/package-json/blob/9d7f8d4306239f258c85755685923d8bda70ee68/lib/index.js#L52

Since `type-fest`'s `PackageJson` type only has optionals, I think it shouldn't matter that we use that type here as I think we can assume it's adhering the correct schema.
https://github.com/sindresorhus/type-fest/blob/main/source/package-json.d.ts

I would still just use the dependency instead of copying it over, as that would mean we would also need to copy over the `LiteralUnion` type (which is used in `PackageJson` type) & `Primitive` type (which is used in `LiteralUnion` type), which creates even more maintenance work in this repo.
https://github.com/sindresorhus/type-fest/blob/c5796f5fce6fc8346792929468159648caec30e0/source/package-json.d.ts#L213-L224
https://github.com/sindresorhus/type-fest/blob/c5796f5fce6fc8346792929468159648caec30e0/source/literal-union.d.ts#L34
https://github.com/sindresorhus/type-fest/blob/c5796f5fce6fc8346792929468159648caec30e0/source/primitive.d.ts

**Comment by andrewbranch** [comment]
type-fest doesn’t even check without `"lib": "es2020"`. That’s not a reasonable constraint to put on every consumer of these types. The type is more complicated than it needs to be; you’re free to decide how many details like the `LiteralUnion` strategy you want to include.

**Comment by MichaelDeBoey** [comment]
@andrewbranch I included the `type-fest` files into this PR

File: types/npmcli__package-json/index.d.ts [hunk 1]
```diff
@@ -0,0 +1,21 @@
...
+// Definitions: https://github.com/DefinitelyTyped/DefinitelyTyped
+
+import type { PackageJson } from './package-json';
+
+export type Content = PackageJson;
+
+declare class NPMCliPackageJson {
+    constructor(path: string);
+
+    static load: (path?: string) => Promise<NPMCliPackageJson>;
```
Comment (peterblazejewicz) [inline types/npmcli__package-json/index.d.ts:13]:
```suggestion
    /** @async */
    static load: (path?: string) => Promise<NPMCliPackageJson>;
```
as the `.d.ts` cannot contain `async` keyword:
https://jsdoc.app/tags-async.html

Comment (peterblazejewicz) [reply types/npmcli__package-json/index.d.ts:13]: and there should be an instance method returning same results

File: types/npmcli__package-json/index.d.ts [hunk 2]
```diff
@@ -0,0 +1,21 @@
...
+    save: () => Promise<void>;
+    update: (content: Content) => void;
+
+    content: Content;
+}
+
+export type { PackageJson } from './package-json';
+export default NPMCliPackageJson;
```
Comment (peterblazejewicz) [inline types/npmcli__package-json/index.d.ts:21]: at the moment, IMO this is proper CJS module:
https://github.com/npm/package-json/blob/6007cad20f83e2dbc80d3453abba3c01a660df83/lib/index.js#L109
so that would require:
```suggestion
export = PackageJson;
```

File: types/npmcli__package-json/tslint.json [hunk 3]
```diff
@@ -0,0 +1,6 @@
+{
+    "extends": "@definitelytyped/dtslint/dt.json",
+    "rules": {
+        "npm-naming": false
+    }
```
Comment (peterblazejewicz) [inline types/npmcli__package-json/tslint.json:5]:
```suggestion
    "extends": "@definitelytyped/dtslint/dt.json"
```
why you would need this rule exclusion? It looks perfect NPM kind of package, resolvable via NPM

File: types/npmcli__package-json/index.d.ts (hunk 2 above)
Comment (MichaelDeBoey) [reply types/npmcli__package-json/index.d.ts:21]: That's not possible according to the compiler

File: types/npmcli__package-json/index.d.ts (hunk 1 above)
Comment (MichaelDeBoey) [reply types/npmcli__package-json/index.d.ts:13]: Done!

File: types/npmcli__package-json/tslint.json (hunk 3 above)
Comment (MichaelDeBoey) [reply types/npmcli__package-json/tslint.json:5]: Without this, I CI wouldn't pass

https://github.com/DefinitelyTyped/DefinitelyTyped/runs/6330895335?check_suite_focus=true

File: types/npmcli__package-json/index.d.ts (hunk 2 above)
Comment (peterblazejewicz) [reply types/npmcli__package-json/index.d.ts:22]: that's OK, you would have to either use a symbol to export everything (excluding types) or use a namespace. I would try this:
```ts
import type { PackageJson as PackageJsonType } from './package-json';

declare class PackageJson {
    static load: (path?: string) => Promise<PackageJson>;
    constructor(path: string);

    save: () => Promise<void>;
    update: (content: PackageJson.Content) => void;

    content: PackageJson.Content;
}

declare namespace PackageJson {
    type Content = PackageJsonType;
}

export = PackageJson;
```
(please rename class to a proper name used in the module)

File: types/npmcli__package-json/tslint.json (hunk 3 above)
Comment (peterblazejewicz) [reply types/npmcli__package-json/tslint.json:5]: it's about the error I've mentoned, this is a CJS module:
```txt
ERROR: 1:1   npm-naming  The declaration doesn't match the JavaScript module 'npmcli__package-json'. Reason:
The declaration should use 'export =' syntax because the JavaScript source uses 'module.exports =' syntax and 'module.exports' can be called or constructed.
```
once fixed, this exclusionn won't be needed I think

File: types/npmcli__package-json/index.d.ts (hunk 2 above)
Comment (MichaelDeBoey) [reply types/npmcli__package-json/index.d.ts:22]: Changing the code to

```ts
import type { PackageJson as PackageJsonType } from './package-json';

declare class NPMCliPackageJson {
    constructor(path: string);

    /** @async */
    static load: (path?: string) => Promise<NPMCliPackageJson>;
    save: () => Promise<void>;
    update: (content: NPMCliPackageJson.Content) => void;

    content: NPMCliPackageJson.Content;
}

declare namespace NPMCliPackageJson {
    type Content = PackageJsonType;

    type PackageJson = PackageJsonType;
}

export = NPMCliPackageJson;
```

Would give the following error in the test file

Comment (peterblazejewicz) [reply types/npmcli__package-json/index.d.ts:22]: the error is correct, so please update the rest of the code, in case of imports:
```ts
import ... = require('....');
```
https://www.typescriptlang.org/docs/handbook/modules.html#export--and-import--require
or:
```ts
import * as .... from '...';
```
(also supported).
The DT modules promotes interop between types definitions. When consumed by users, 'esModuleInterop' is probably already set as being default from some time in TS, so this:
```ts
import .... from '...';
```
would work out of the box.
https://github.com/DefinitelyTyped/DefinitelyTyped#esmoduleinteropallowsyntheticdefaultimports

Comment (MichaelDeBoey) [reply types/npmcli__package-json/index.d.ts:22]: @peterblazejewicz That would make the test pass again, but that's not how the package is meant to be used

It's meant to be used like

```ts
import PackageJson from '@npmcli/package-json';
```

Comment (peterblazejewicz) [reply types/npmcli__package-json/index.d.ts:22]: Yes, I understand:

> When consumed by users, 'esModuleInterop' is probably already set as being default from some time in TS, so this:

IMO this module is a classic Node CJS, without default export or default export alias:
https://github.com/npm/package-json/blob/main/lib/index.js#L109
https://www.typescriptlang.org/tsconfig#esModuleInterop

Comment (peterblazejewicz) [reply types/npmcli__package-json/index.d.ts:22]: Let me rephrase. I'm asking to fix definition for purpose of DefinitelyType and TS. 
It will work perfectly OK for users downloading those types.

Comment (MichaelDeBoey) [reply types/npmcli__package-json/index.d.ts:22]: OK I don't think I understand correctly then 🤔

The error I got was because I changed the code to how you suggested it should be.
So what do you think I should do exactly to make it all work?

Comment (peterblazejewicz) [reply types/npmcli__package-json/index.d.ts:22]: it's not me, this is just advised, I could write:
```ts
import type { PackageJson as PackageJsonType } from './package-json';

declare namespace PackageJson {
    type Content = PackageJsonType;
}

declare class PackageJson {
    static load: (path?: string) => Promise<PackageJson>;
    constructor(path: string);

    save: () => Promise<void>;
    update: (content: PackageJson.Content) => void;

    content: PackageJson.Content;
}

export = PackageJson;

// this export class with proper name and re-exports your types via namespace export

// test within DT repo
import PackageJson = require('@npmcli/package-json');

(async () => {
    const pkgJson = await PackageJson.load('./');

    pkgJson.update({
        dependencies: {
            a: '^1.0.0',
            b: '^1.2.3',
        },
        workspaces: ['./new-workspace'],
    });

    await pkgJson.save();
})();

// usage via default TS configs would be
// import PackageJson from '@npmcli/package-json';
// or const PackageJson = require('@npmcli/package-json');
```

Comment (MichaelDeBoey) [reply types/npmcli__package-json/index.d.ts]: If people could still do

```ts
import PackageJson from '@npmcli/package-json';
```

Then necessary changes are made.

**Review by andrewbranch — CHANGES_REQUESTED** [review]
The existence of `literal-union.d.ts` and `primitive.d.ts` and `package-json.d.ts` imply the existence of `.js` files by the same names in the implementation package. I.e., this should be an error

```ts
import {} from "@npmcli/package-json/primitive"
```

because it will crash at runtime, but it is allowed because you have a module file there. Unfortunately you need to inline your utilities into index.d.ts.

I also _strongly_ recommend dropping the LiteralUnion hack. It’s a nonsense type and nobody should be surprised if a future version of TypeScript accidentally breaks it.

**Review by andrewbranch — APPROVED** [review]

**Comment by MichaelDeBoey** [comment]
Ready to merge
